### PR TITLE
Application choice submission errors updated

### PR DESCRIPTION
**Pull request by inulty-dfe**
## Context

The Application review screen for unsubmitted applications shows error messages when the application is not in a submittable state. Currently, all invalid error messages are shown on the page. We are changing this so that we only show the highest priority error message that is preventing the application from being submitted.

1. If the candidate is not eligible to take the course for reasons relating to visas and immigration
2. The course is not available to take applications (outside cycle or course has later application date)
3. The course has no vacancies or has been removed from the service
4. The candidate has not completed their details and they do not need to complete Science GCSE
5. The candidate has completed all their details except Science GCSE
6. The candidate has not completed Science GCSE and at least one other section in details

We validate each of these contexts and only show the error message for the first failed validation.

## Changes proposed in this pull request

Some validators have been removed as they have no purpose since the new validations are introduced.
Add `find_url` to routes
Remove single 'ApplicationChoiceSubmission' spec and instead have separate smaller files for each validation context.

### Scenarios

**1. If the candidate is not eligible to take the course for reasons relating to visas and immigration**

- nationality is not British/Irish
- no right to work in UK
- choose a course that does not sponsor visas

(If you pick a course that does sponsor visas, you can submit.)

**2. The course is not available to take applications (outside cycle or course has later application date)**

- change cycle to before apply opens
or
- change course start date to some time in the future

**3. The Course has no vacancies or has been removed from the service**

- set all course options `vacancy_status: 'no_vacancies'
or
- set `open_on_apply: false`

**4. The candidate has not completed their details and they do not need to complete Science GCSE**

- pick a non-Primary course
- leave any of the details sections incomplete

**5. The candidate has completed all their details except Science GCSE**

- pick a Primary course
- complete all sections of details except Science GCSE

**6. The candidate has not completed Science GCSE and at least one other section in details**

- pick Primary course
- leave Science GCSE incomplete and at least one other details section

## Link to Trello card

[Trello Ticket ](https://trello.com/c/UQxutEfA/544-apply-show-content-when-we-block-candidates-from-submitting-in-a-better-way)

### CODE REVIEW

**Comment by inulty-dfe** [comment]
[Apply: Show content when we block candidates from submitting in a better way](https://trello.com/c/UQxutEfA/544-apply-show-content-when-we-block-candidates-from-submitting-in-a-better-way)

**Comment by inulty-dfe** [comment]
[Apply: Show content when we block candidates from submitting in a better way](https://trello.com/c/UQxutEfA/544-apply-show-content-when-we-block-candidates-from-submitting-in-a-better-way)

File: app/presenters/candidate_interface/application_form_presenter.rb [hunk 1]
```diff
@@ -36,7 +36,7 @@ def sections_with_validations
     def sections_with_completion
       [
         # "Courses" section
-        [:course_choices, course_choices_completed?],
+        ([:course_choices, course_choices_completed?] if RecruitmentCycle.current_year < 2024),
```
Comment (JR-G) [inline app/presenters/candidate_interface/application_form_presenter.rb:39]: Minor as we'll look to remove this field but I think there's a `CONTINUOUS_APPLICATIONS_CYCLE_YEAR` constant

File: spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb [hunk 2]
```diff
@@ -47,6 +47,8 @@
     when_i_have_completed_my_application_and_added_primary_as_course_choice
     when_i_have_not_completed_science_gcse
     and_i_continue_with_my_application
+
+    # ERROR
```
Comment (inulty-dfe) [inline spec/system/candidate_interface/continuous_applications/submitting/candidate_submitting_application_spec.rb]: :space_invader:

**Review by MaeveRoseveare — APPROVED** [review]
Amazing stuff. The only one I wasn't able to simulate in the review app is the one we show in between cycles. But the content in the file is correct!

File: app/validators/immigration_status_validator.rb [hunk 3]
```diff
@@ -20,4 +16,25 @@ def course_can_sponsor_visa?(application_choice)
...
+  def view
+    Class.new do
+      include ActionView::Helpers::UrlHelper
+      include GovukLinkHelper
+    end.new
+  end
+
+  def url_helpers
+    Rails.application.routes.url_helpers
+  end
```
Comment (JR-G) [inline app/validators/immigration_status_validator.rb:35]: What are your thoughts on using `include Rails.application.routes.url_helpers` within these validators. The others don't appear to have the helper method?

File: app/components/candidate_interface/continuous_applications/application_submit_component.html.erb [hunk 4]
```diff
@@ -10,21 +10,14 @@
   <%= form.govuk_submit t('continue') %>
 <% else %>
   <% errors.each do |error| %>
-    <p class="govuk-body">
-      <% if error.type == :incomplete_details %>
-        <p class="govuk-body">You need to <%= govuk_link_to('complete your details', candidate_interface_continuous_applications_details_path) %> before you can submit this application.</p>
-        <p class="govuk-body">This application will be saved as a draft while you finish your details.</p>
-      <% elsif error.type == :immigration_status %>
+    <% error.message.split(/\n+/).each_with_index do |line, idx| %>
+      <% if idx == 0 && error.type == :immigration_status %>
```
Comment (tomas-stefano) [inline app/components/candidate_interface/continuous_applications/application_submit_component.html.erb:14]: Minor but can we renamed the var to `index`, please? and `index.zero?`

File: app/services/candidate_interface/continuous_applications/application_choice_submission.rb [hunk 5]
```diff
@@ -6,13 +6,19 @@ class ApplicationChoiceSubmission
...
+                course_unavailable: { if: :validate_choice? },
+                incomplete_details: { if: :validate_choice? },
+                incomplete_primary_course_details: { if: :validate_choice? },
+                incomplete_including_primary_course_details: { if: :validate_choice? },
+                can_add_more_choices: true
+
+    private
+
+      def validate_choice?
+        errors.exclude?(:application_choice)
```
Comment (tomas-stefano) [inline app/services/candidate_interface/continuous_applications/application_choice_submission.rb:20]: An interesting approach, so the business object only will hold one of the errors always? 🤔

File: app/validators/course_unavailable_validator.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+class CourseUnavailableValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, application_choice)
+    course = application_choice.current_course
+
+    return if !course.full? &&
```
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/validators/course_unavailable_validator.rb:5]: Oh that packages `site_still_valid: true` and `vacancy_status: 'vacancies'`.

File: app/validators/immigration_status_validator.rb (hunk 3 above)
Comment (inulty-dfe) [reply app/validators/immigration_status_validator.rb:35]: There are too many methods in `url_helpers` to include them all in a single class. This class is only using one of them.

Including the whole module would flood the object with methods it doesn't need.

File: app/components/candidate_interface/continuous_applications/application_submit_component.html.erb (hunk 4 above)
Comment (inulty-dfe) [reply app/components/candidate_interface/continuous_applications/application_submit_component.html.erb:14]: Sure

File: app/services/candidate_interface/continuous_applications/application_choice_submission.rb (hunk 5 above)
Comment (inulty-dfe) [reply app/services/candidate_interface/continuous_applications/application_choice_submission.rb:20]: Yeah, I'm confident that the validations are called in the order they're defined.

File: app/presenters/candidate_interface/application_form_presenter.rb (hunk 1 above)
Comment (inulty-dfe) [reply app/presenters/candidate_interface/application_form_presenter.rb:39]: Yeah I didn't want to get into this class. It looks like it needs revisiting. I can add the const

File: app/validators/course_unavailable_validator.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (inulty-dfe) [reply app/validators/course_unavailable_validator.rb:5]:
```
  # Course
  def available?
    course_options.available.present?
  end
```

```
  # CourseOption

  scope :selectable, -> { where(site_still_valid: true) }

  scope :available, lambda {
    selectable.where(vacancy_status: 'vacancies')
  }
```

`course.available?` checks

> Do any of the course options have vacancies and an available site.

The current implementation checks

> Do any of the course options have vacancies and does the chosen course option have an available site.

**Review by JR-G — APPROVED** [review]

**Pull request merged**